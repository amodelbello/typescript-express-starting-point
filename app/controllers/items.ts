import { Request, Response } from 'express'
import { Item } from '../types'

export const getItems = (req: Request, res: Response): void => {
  const item: Item = {
    id: 'id-123',
    name: 'Name of the item',
    description: 'Description of the item',
  }
  res.send([item])
}

export const getItem = (req: Request, res: Response): void => {
  const itemId = req.params.itemId
  const item: Item = {
    id: itemId,
    name: `Name of item: ${itemId}`,
    description: `Description of item: ${itemId}`,
  }
  res.send(item)
}

export const addItem = (req: Request, res: Response): void => {
  const item: Item = {
    id: 'added-item-123',
    ...req.body,
  }
  res.send(item)
}

export const updateItem = (req: Request, res: Response): void => {
  const itemId = req.params.itemId
  const item: Item = {
    id: itemId,
    ...req.body,
  }
  res.send(item)
}

export const deleteItem = (req: Request, res: Response): void => {
  const itemId = req.params.itemId
  const item: Item = {
    id: itemId,
    name: `Name of deleted item: ${itemId}`,
    description: `Description of deleted item: ${itemId}`,
  }
  res.send(item)
}
