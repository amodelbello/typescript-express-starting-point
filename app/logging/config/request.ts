import dotenv from 'dotenv'
import * as colors from '../colors'
import { TransformableInfo } from 'logform'
import { format, transports } from 'winston'
dotenv.config()
const { combine, timestamp, label, printf } = format

const config = {
  level: process.env.LOG_LEVEL /* istanbul ignore next */ || 'info',
  format: combine(
    format(
      (info): TransformableInfo => {
        info.level = info.level.toLocaleUpperCase()
        return info
      },
    )(),
    combine(
      label({
        label: 'Request',
      }),
      timestamp({
        format: 'YYYY-MM-DD h:mm:SS A',
      }),
      printf(({ label, timestamp, level, message }): string => {
        return `${timestamp} [${colors.gray(label)}] ${colors.green(level)} : ${colors.gray(message)}`
      }),
    ),
  ),
  transports: [
    new transports.Console({
      silent: process.env.NODE_ENV === 'testing' && process.env.LOG_OUTPUT_WHILE_RUNNING_TESTS !== 'true',
    }),
  ],
}
export default config
