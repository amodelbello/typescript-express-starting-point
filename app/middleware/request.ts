import { Request, Response, NextFunction } from 'express'
import { v4 as uuidv4 } from 'uuid'

export const addFingerprintToRequest = (req: Request, res: Response, next: NextFunction): void => {
  req.id = uuidv4()
  next()
}
