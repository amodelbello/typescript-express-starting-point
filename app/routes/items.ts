import express from 'express'
import { getItems, getItem, addItem, updateItem, deleteItem } from '../controllers/items'

const router = express.Router()
router
  .route('/')
  .get(getItems)
  .post(addItem)

router
  .route('/:itemId')
  .get(getItem)
  .put(updateItem)
  .delete(deleteItem)

export default router
