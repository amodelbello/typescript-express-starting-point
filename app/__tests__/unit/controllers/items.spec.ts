import request from 'supertest'
import ExpressServer from '../../../app'

import { id as itemId, item as testItem } from '../../__fixtures__/items'

const server = new ExpressServer(4003)
const { app } = server

beforeEach(async () => {
  await server.run(true)
})

afterEach(async () => {
  console.log('items.spec.ts server stop')
  await server.stop()
})

describe('Items', () => {
  it('GET /items`', async () => {
    await request(app)
      .get('/items')
      .expect(200)
      .then(({ body }) => {
        expect(body).toMatchSnapshot()
      })
  })

  it('GET /items/:id`', async () => {
    await request(app)
      .get(`/items/${itemId}`)
      .expect(200)
      .then(({ body }) => {
        expect(body).toMatchSnapshot()
      })
  })

  it('POST /items`', async () => {
    await request(app)
      .post('/items')
      .send({ name: testItem.name, description: testItem.description })
      .expect(200)
      .then(({ body }) => {
        expect(body).toMatchSnapshot()
      })
  })

  it('PUT /items/:id`', async () => {
    await request(app)
      .put(`/items/${itemId}`)
      .send(testItem)
      .expect(200)
      .then(({ body }) => {
        expect(body).toMatchSnapshot()
      })
  })

  it('DELETE /items/:id`', async () => {
    await request(app)
      .delete(`/items/${itemId}`)
      .expect(200)
      .then(({ body }) => {
        expect(body).toMatchSnapshot()
      })
  })
})
