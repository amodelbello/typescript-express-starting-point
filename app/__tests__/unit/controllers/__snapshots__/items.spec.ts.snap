// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items DELETE /items/:id\` 1`] = `
Object {
  "description": "Description of deleted item: this-is-a-test-item-id-1234",
  "id": "this-is-a-test-item-id-1234",
  "name": "Name of deleted item: this-is-a-test-item-id-1234",
}
`;

exports[`Items GET /items/:id\` 1`] = `
Object {
  "description": "Description of item: this-is-a-test-item-id-1234",
  "id": "this-is-a-test-item-id-1234",
  "name": "Name of item: this-is-a-test-item-id-1234",
}
`;

exports[`Items GET /items\` 1`] = `
Array [
  Object {
    "description": "Description of the item",
    "id": "id-123",
    "name": "Name of the item",
  },
]
`;

exports[`Items POST /items\` 1`] = `
Object {
  "description": "Description of item: this-is-a-test-item-id-1234",
  "id": "added-item-123",
  "name": "Name of item: this-is-a-test-item-id-1234",
}
`;

exports[`Items PUT /items/:id\` 1`] = `
Object {
  "description": "Description of item: this-is-a-test-item-id-1234",
  "id": "this-is-a-test-item-id-1234",
  "name": "Name of item: this-is-a-test-item-id-1234",
}
`;
